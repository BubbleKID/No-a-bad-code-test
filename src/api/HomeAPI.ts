export const ITEM_API_URL = 'http://prototype.dev.carter.digital/fed-test/items.json';