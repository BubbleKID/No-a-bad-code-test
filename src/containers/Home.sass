.main-container
    margin: auto
    max-width: 1200px
.card-container 
    display: flex
    flex-wrap: wrap
    justify-content: center
