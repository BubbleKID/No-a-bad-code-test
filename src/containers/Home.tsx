import React, { useState, useEffect } from 'react';
import { ITEM_API_URL } from '../api/HomeAPI';
import Card from '../components/Card/Card';
import './Home.sass';

export default function Home() {
    const [items, setItems] = useState([] as any[]);
    
    useEffect(() => {
        const fetchItems = async () => {
            try {
                const response = await fetch(ITEM_API_URL)
                const data = await response.json();
                setItems(data.items);
            } catch (err) {
                console.log(err);
            }
        }
        fetchItems();
    }, []);
    
    return(
        <div className="main-container">
            <div className="card-container">
                { items.map(item => <Card key={item.id} {...item} height={280} width={180}/>) }
            </div>
        </div>
    );
}