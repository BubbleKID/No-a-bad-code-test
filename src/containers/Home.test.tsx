import React from 'react';
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import TestRenderer  from 'react-test-renderer';
import Home from './Home';
import { ITEM_API_URL } from '../api/HomeAPI';

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

test('Home should be rendered', () => {
    const component = TestRenderer.create(<Home/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("should fetch data", async () => {
    const fakeItems = {
        "items": [
            {
                "id": "1",
                "title": "title1",
                "description": "description1",
                "category": "Programs, Company",
                "link": "http://www.google.com.au",
                "featured": "true"
            },
            {
                "id": "2",
                "title": "title2",
                "category": "description2",
                "link": "http://www.google.com.au",
                "featured": "false"
            },
        ]
    };
    const promiseMock = new Promise<any>((resolve) => {
        resolve({
          json: () => Promise.resolve(fakeItems)
        })
    });

    global.fetch = jest.fn().mockImplementation(() => promiseMock);
  
    await act(async () => {
      render(<Home />, container);
    });
  
    // first item
    expect(container.querySelectorAll(".card-title")[0].textContent).toBe(fakeItems.items[0].title);
    expect(container.querySelectorAll(".description")[0].textContent).toBe(fakeItems.items[0].description);
    expect(container.querySelectorAll(".featured .card-title")[0].textContent).toBe(fakeItems.items[0].title);
    // second item
    expect(container.querySelectorAll(".card-title")[1].textContent).toBe(fakeItems.items[1].title);

    delete global.fetch
});

test("should return error message when fetch is rejected", async () => {
    const promiseMock = new Promise<any>((resolve, reject) => {
        reject('fake error')
    });

    global.fetch = jest.fn().mockImplementation(() => promiseMock);
  
    await act(async () => {
      render(<Home />, container);
    });

    expect(fetch).toHaveBeenCalledWith(ITEM_API_URL);
    expect(fetch).rejects.toMatch('fake error');
    delete global.fetch
});