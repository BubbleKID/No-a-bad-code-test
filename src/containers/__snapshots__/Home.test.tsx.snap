// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should be rendered 1`] = `
<div
  className="card-container"
/>
`;
